unifi: embedded libsnappyjava.so has unresolvable dynamic link to libstdc++.so

Version: 16.03pre75740.f727f81 (Emu)

To reproduce:

1. Try to run set-inform on a gateway device to point it to the nixos host running unifi.
2. The device will fail to be adopted, repeatedly.
3. systemctl status -l unifi will show:

        Feb 09 19:07:29 unifi java[10815]: at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        Feb 09 19:07:29 unifi java[10815]: at java.lang.Thread.run(Thread.java:745)
        Feb 09 19:07:29 unifi java[10815]: Caused by: java.lang.UnsatisfiedLinkError: /tmp/snappy-1.0.5-libsnappyjava.so: libstdc++.so.6: cannot open shared object file: No such file or directory
        Feb 09 19:07:29 unifi java[10815]: at java.lang.ClassLoader$NativeLibrary.load(Native Method)
        Feb 09 19:07:29 unifi java[10815]: at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1938)
        Feb 09 19:07:29 unifi java[10815]: at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1821)
        Feb 09 19:07:29 unifi java[10815]: at java.lang.Runtime.load0(Runtime.java:809)
        Feb 09 19:07:29 unifi java[10815]: at java.lang.System.load(System.java:1086)
        Feb 09 19:07:29 unifi java[10815]: at org.xerial.snappy.SnappyNativeLoader.load(SnappyNativeLoader.java:39)
        Feb 09 19:07:29 unifi java[10815]: ... 31 more



This is because the java service extracts an embedded snappy-1.0.5-libsnappyjava.so to /tmp and tries to use it. It fails, since this is a binary compiled without nixos. Since this package is built from a tarball of binaries, I don't know if it's even possible to remedy this cleanly. My workaround is adding the following to configuration.nix:

    # HACK to fix /tmp/snappy-1.0.5-libsnappyjava.so (extracted by service from jar) not having link to any libstdc++.so.6
    environment.systemPackages = with pkgs; [ gcc ];
    systemd.services.unifi.environment= { LD_LIBRARY_PATH = "/run/current-system/sw/lib"; };

This allows the java parent process to find libstdc++.so.6 at runtime. I don't know any better way to handle this requirement; hopefully a more experienced contributor knows a better solution.

--- a/nixos/modules/services/networking/unifi.nix
+++ b/nixos/modules/services/networking/unifi.nix
@@ -61,6 +61,8 @@ in
       partOf = systemdMountPoints;
       bindsTo = systemdMountPoints;
       unitConfig.RequiresMountsFor = stateDir;
+      # This a HACK to fix missing dependencies of dynamic libs extracted from jars
+      environment.LD_LIBRARY_PATH = with pkgs.stdenv; "${cc.cc}/lib";
 
       preStart = ''
         # Ensure privacy of state

